src/test/memory/test_mmap.rs

```diff
@@ -0,0 +1,434 @@
+unsafe fn mmap_using_mmap(
+    addr: Option<&mut [u8]>,
+    prot: i32,
+    flags: i32,
+    fd: i32,
+    offset: i64,
+) -> *mut libc::c_void {
+    let (ptr, len) = if addr.is_some() {
+        let addr = addr.unwrap();
+        (addr.as_mut_ptr() as *mut libc::c_void, addr.len())
+    } else {
+        (0 as *mut libc::c_void, MAPLEN)
+    };
+
+    libc::mmap(
+        ptr,
+        len,
+        prot as libc::c_int,
+        flags as libc::c_int,
+        fd as libc::c_int,
+        offset as libc::off_t,
+    )
+}
+
+struct TestMmapFileData {
+    mmap_fn: unsafe fn(Option<&mut [u8]>, i32, i32, i32, i64) -> *mut libc::c_void,
+    offset: libc::off_t,
+}
+
+fn test_mmap_file(data: &TestMmapFileData) -> Result<(), String> {
+    let mut template = "test_mmapXXXXXX".to_string();
+
+    /* Get a file that we can mmap and write into. */
+    let temp_fd = unsafe { libc::mkstemp(template.as_mut_ptr() as *mut libc::c_char) };
+    test_utils::assert_true_and_errno(temp_fd != -1);
+
+    let fp = unsafe { libc::fdopen(temp_fd, "w+".as_ptr() as *const libc::c_char) };
+
+    /* Make sure there is enough space to write after the mmap. */
+    {
+        let rv =
+            unsafe { libc::posix_fallocate(temp_fd as libc::c_int, data.offset, MAPLEN as i64) };
+        assert_eq!(rv, 0, "{}", test_utils::get_errno_message(rv));
+    }
+
+    /* Init a msg to write. */
+    let mut msg = [0_i8; MAPLEN];
+    unsafe {
+        libc::snprintf(
+            msg.as_mut_ptr() as *mut libc::c_char,
+            MAPLEN,
+            "Hello world!".as_ptr() as *const i8,
+        )
+    };
+    test_utils::assert_errno_nonneg();
+
+    /* Do the mmap and write the message into the resulting mem location. */
+    let mapbuf = unsafe {
+        (data.mmap_fn)(
+            None,
+            libc::PROT_READ | libc::PROT_WRITE,
+            libc::MAP_SHARED,
+            temp_fd,
+            data.offset as i64,
+        )
+    };
+
+    assert!(mapbuf as i32 != libc::MAP_FAILED as i32);
+
+    unsafe { libc::snprintf(mapbuf as *mut i8, MAPLEN, msg.as_ptr()) };
+    test_utils::assert_errno_nonneg();
+
+    unsafe { libc::munmap(mapbuf, MAPLEN) };
+    test_utils::assert_errno_nonneg();
+
+    unsafe { libc::fseek(fp, data.offset, libc::SEEK_SET) };
+    test_utils::assert_errno_nonneg();
+
+    let mut rdbuf = [0_i8; MAPLEN];
+    let rdret = unsafe { libc::fread(rdbuf.as_mut_ptr() as *mut libc::c_void, 1, MAPLEN, fp) };
+    assert!(rdret > 0);
+
+    assert_eq!(msg[..], rdbuf[..]);
+
+    unsafe { libc::fclose(fp) };
+    test_utils::assert_errno_nonneg();
+
+    Ok(())
+}
+
+fn page_size() -> libc::size_t {
+    unsafe { libc::sysconf(libc::_SC_PAGE_SIZE) as libc::size_t }
+}
+
+fn init_buf(buf: *mut libc::c_void, size: usize) {
+    unsafe {
+        let b = std::slice::from_raw_parts_mut::<u8>(buf as *mut u8, size);
+        for (i, byte) in b.iter_mut().enumerate() {
+            *byte = (i / page_size()) as u8;
+        }
+    }
+}
```

Comment:
Converted to take a `&mut [u8]`. However, `mmap_and_init_buf` still needs to return a `*mut libc::c_void` because of subsequent uses of `mremap` and `munmap`. Trying to keep the return as either `&'static mut [u8]` or `Vec<u8>` results in undefined behavior and segfaults :scream: 

`Vec` has a `into_raw_parts` nightly API, but it hardly seems worth the hassle of forcing a nightly compiler (especially for a single test).